### PR TITLE
Test of CustomScrollViewExampleApp

### PR DESCRIPTION
**Pull request by miechoo**
CustomScrollViewExampleApp 
Part of https://github.com/flutter/flutter/issues/130459

The test is checking: 
- if all crucial Widgets are initially visible
- if IconButton click will expand existing SliverList
- if IconButton click and mouse scroll will reveal additional SliverList

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by nate-thegrate — COMMENTED** [review]
Could you take a look at the failing check? (see [how to fix failing checks](https://github.com/nate-thegrate/flutter/blob/test-fail-wiki-page/docs/contributing/testing/Test-failure-troubleshooting.md) for more info)

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:87]: Thanks for adding in the additional `reason`, I don't see that being done very often!

I would recommend getting rid of the line break (here and below), based on one of the [style guidelines](https://github.com/flutter/flutter/blob/05f294c7116f30a93195b479d55539f457edd83f/docs/contributing/Style-guide-for-Flutter-repo.md#prefer-a-maximum-line-length-of-80-characters):

```suggestion
      reason: 'Expected one, initial (bottom), SliverList in the CustomScrollView',
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:76]: Having both of the testWidgets arguments start on the same line, with a line break in between calls, would be ideal for consistency with other tests in the framework.

```suggestion

  testWidgets('IconButton click extends existing SliverList', (WidgetTester tester) async {
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:20]: Comments like this are nice, but they also carry a risk: if someone updates the test in the future but forgets to change the comments, things could get very confusing.

Since each `expect()` call already has a `reason`, I think it'd be best to remove the comments here.

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:184]: Not a huge deal whatsoever, but maybe consider making a whitespace change for readability:

```suggestion
    await tester.sendEventToBinding(
      PointerScrollEvent(position: location, scrollDelta: const Offset(0, -1)),
    );
```

**Review by nate-thegrate — COMMENTED** [review]
Overall, this looks phenomenal—thanks for the high-quality contribution!

I made a few small suggestions (and it looks like one of the checks might be stuck; hopefully it'll resolve itself once some changes are pushed).

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:121]: nit: add a blank line here

```suggestion
  });

```

**Review by nate-thegrate — APPROVED** [review]
LGTM, thanks for this contribution!

(We'll be able to merge once the PR has 2 approvals.)

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:155]:
```suggestion
    // It should not reveal additional (top) SliverList
```

Apparently, [Linux Analyze](https://github.com/flutter/flutter/pull/152431/checks?check_run_id=28241547944) is upset about this 😒

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:156]: nit: grammar

```suggestion
    // because the additional SliverList does not exist yet.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:18]:
```suggestion
    expect(find.byType(Scaffold), findsOne);
```

Adding 'reason' is not bad in itself but I think we should reserved it to non-obvious cases because:
- it makes the code harder to read by splitting it on several lines.
- it took more effort to write (and might lead to typos).
- it takes more effort to read and then realized that this sentence just restate the code.
- it takes more effort to review (looking for typos, formatting issues, understanding if the comment is right).

This is similar to over-commenting: it starts from a good intention but leads to readibility and maintenance issues.
Reason texts and comments are welcome when they add value.

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:96]:
```suggestion
    // Tap the IconButton in the AppBar.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:120]:
```suggestion
    expect(find.widgetWithText(CustomScrollView, 'Item: 1'));
```

`widgetWithText` is a nice utility that helps to reduce code (and improve readability).
Consider applying it to all similar cases in this file.

**Review by bleroux — COMMENTED** [review]
Thanks for the contribution!
The quality shows that you put effort into crafting it. :pray: 

My main feeback is that theses tests are too verbose which lead to maintenance and reviewing problem.
See my specific comments to reduce the verbosity.

**Review by bleroux — COMMENTED** [review]
There are many ways to write tests. In that particular context, testing Flutter API examples, the main goals are consistency and simplicity.

I encourage you to have a look at the other tests for API examples (to get an oversight of the style, the formatting, etc). Be aware that the consistency is not always there because there were many contributors and many reviewers over the time.

Using groups might make sense for a very long test file (some test files in the Flutter repo are more than 10 thousands lines :sweat_smile: ).
For tests related to API examples, groups are not needed because those test files usually contain just a couple of tests (and often just one). The prefer way is to:
- Write good test descriptions.
- When needed, add some comments to separate sections inside tests.
- Avoid repetitions.
- Do not rely on ‘reason’ for documentation, prefer comments (see my specific comment on the review).

For this particular example, I think only two tests are needed:
- One for the initial state (is the app bar title there, is the plus button there, is ‘Item 0’ visible).
- Another one for the state after the tap (is ‘item 0’ and ‘item 1’ are visible, and is ‘item -1’ not visible). In that same test you can terminate with a scroll gesture to check that all items are visible (a comment before that section would be welcome).

Another small advice: try to use only one GitHub account when working on a PR. Using several GitHub accounts might be a problem with the CLA check and it might be confusing for others.

**Comment by nate-thegrate** [comment]
> Why the Mac_x64 native_ui_tests_macos test fails?

It was just a flake, so no worries!

If this happens in the future, you can wait for somebody with commit access to re-run the check, or you can trigger all the checks by pushing an update (e.g. updating with the latest changes to `master`).

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:20]: Super tiny whitespace nit (per the [style guide](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation)):

```dart
    // this would work
    expect(find.descendant(
      of: find.byType(Scaffold),
      matching: find.byType(AppBar),
    ), findsOne);


    // here's another option
    expect(
      find.descendant(
        of: find.byType(Scaffold),
        matching: find.byType(AppBar),
      ),
      findsOne,
    );
```

Comment (nate-thegrate) [reply examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart]: It looks like these were changed to:

```dart
    expect(find.descendant(
          of: find.byType(Scaffold),
          matching: find.byType(AppBar),
        ),
        findsOne);
```

which is much better in terms of following the style guide! But it still doesn't quite match the whitespace conventions found elsewhere in the framework.

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:30]: nit:

```suggestion
    expect(find.descendant(
      of: find.byType(Scaffold),
      matching: find.byType(AppBar),
    ), findsOne);
    expect(find.descendant(
      of: find.byType(AppBar),
      matching: find.byType(IconButton),
    ), findsOne);
    expect(find.descendant(
      of: find.byType(Scaffold),
      matching: find.byType(CustomScrollView),
    ), findsOne);
    expect(find.descendant(
      of: find.byType(CustomScrollView),
      matching: find.byType(SliverList),
    ), findsOne);
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:11]:
```suggestion
  testWidgets('What should be visible in the initial state.', (WidgetTester tester) async {
```
Try to keep it simple, it's not easy to find good test description.

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:30]: Thoses checks can be removed because there are checking the structure of the implementation. The goal for thoses tests is to check if the example works from a user point of view.

This can be replace by only one expect that checks if the IconButton with the plus sign is visible.
Eventually we can keep `expect(find.byType(SliverList), findsOne);` (this one might be interesting from a developper point of view, as it emphasizes the solution illustrated by this example).

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:43]: Before scrolling, we should check if 'item 1' is visible and 'item -1' not visible.

**Review by bleroux — COMMENTED** [review]
Making good progress there. See my comment for more simplification and also add some checks before the scroll.

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:17]:
```suggestion
    expect(find.descendant(
      of: find.byType(IconButton),
      matching: find.byIcon(Icons.add),
    ), findsOne);
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:19]:
```suggestion

    // Initial state should present only "Item: 0" on the SliverList.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:25]:
```suggestion
  testWidgets('Items are added correctly', (WidgetTester tester) async {
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:31]:
```suggestion
    // 'Item: -1' is invisible before scrolling.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:35]:
```suggestion
    // Scroll the updated screen.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:47]:
```suggestion
    // An additional SliverList appears.
```

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (bleroux) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:53]:
```suggestion
    // All items are visible.
    expect(find.widgetWithText(SliverList, 'Item: -1'), findsOne);
    expect(find.widgetWithText(SliverList, 'Item: 0'), findsOne);
    expect(find.widgetWithText(SliverList, 'Item: 1'), findsOne);
```

**Review by bleroux — COMMENTED** [review]
Some more cleanup.

**Comment by nate-thegrate** [comment]
In the future, feel free to click "add to batch" on the suggestions, so they can be applied as 1 commit; I believe this needs to be done from the [**Files changed**](https://github.com/flutter/flutter/pull/152431/files) section. Each individual commit triggers all the checks to re-run and sends a notification to the people involved with the PR (not a huge deal though).

File: examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart
Comment (nate-thegrate) [inline examples/api/test/widgets/scroll_view/custom_scroll_view.1_test.dart:5]:
```suggestion
```

```
info • The import of 'package:flutter/gestures.dart' is unnecessary because all of the
       used elements are also provided by the import of 'package:flutter/material.dart'
```

**Review by bleroux — APPROVED** [review]
LGTM! Thanks for your contribution.

**Pull request merged**
